### PR TITLE
8279668: x86: AVX2 versions of vpxor should be asserted

### PR DESCRIPTION
**Pull request by shipilev**
Got the SIGILLs on some machines when testing JDK-8279621. That patch started using shorter `vpxor` versions on `UseAVX = 1` path, which tried to use `VEX.256`-encoded `vpxor` instruction that is only available on AVX2. This should be at very least asserted in assembler code. 

Also fixed the implicit `bool` -> `int` conversion to `vector_len`.

Additional testing:
 - [x] Linux x86_64 fastdebug `tier1` with `-XX:UseAVX=1`

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8279668](https://bugs.openjdk.java.net/browse/JDK-8279668): x86: AVX2 versions of vpxor should be asserted

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Jie Fu](https://openjdk.java.net/census#jiefu) (@DamonFool - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7005/head:pull/7005` \
`$ git checkout pull/7005`

Update a local copy of the PR: \
`$ git checkout pull/7005` \
`$ git pull https://git.openjdk.java.net/jdk pull/7005/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7005`

View PR using the GUI difftool: \
`$ git pr show -t 7005`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7005.diff">https://git.openjdk.java.net/jdk/pull/7005.diff</a>

</details>

### CODE REVIEW

**Comment by merykitty** [comment]
Hi, I think the assert should be inserted in the assembler instead. Also, I would prefer `AVX_256bit` instead of `1`.
Thanks.

**Comment by shipilev** [comment]
> Hi, I think the assert should be inserted in the assembler instead. 

I think `Assembler::vpxor` actually does it right already, and the problem is only in "shortcut" macro-method for AVX2. So the assert is where it should be.

> Also, I would prefer `AVX_256bit` instead of `1`. Thanks.

Good idea, fixed.

**Comment by merykitty** [comment]
What I mean is an assert similar to:
https://github.com/openjdk/jdk/blob/e52bdf0a10d05aee93783d8d4e3a3b349cf4c174/src/hotspot/cpu/x86/assembler_x86.cpp#L2032
Thanks.

**Comment by shipilev** [comment]
> What I mean is an assert similar to:
> https://github.com/openjdk/jdk/blob/e52bdf0a10d05aee93783d8d4e3a3b349cf4c174/src/hotspot/cpu/x86/assembler_x86.cpp#L2032

Ah, I see! Added in new commit.

File: src/hotspot/cpu/x86/macroAssembler_x86.hpp [hunk 1]
```diff
@@ -1489,8 +1489,14 @@ class MacroAssembler: public Assembler {
   void vpxor(XMMRegister dst, XMMRegister nds, AddressLiteral src, int vector_len, Register scratch_reg = rscratch1);
 
   // Simple version for AVX2 256bit vectors
-  void vpxor(XMMRegister dst, XMMRegister src) { Assembler::vpxor(dst, dst, src, true); }
-  void vpxor(XMMRegister dst, Address src) { Assembler::vpxor(dst, dst, src, true); }
+  void vpxor(XMMRegister dst, XMMRegister src) {
+    assert(UseAVX >= 2, "Only with AVX2");
```
Comment (DamonFool) [inline src/hotspot/cpu/x86/macroAssembler_x86.hpp:1493]: I would suggest the assert msg as "Only with AVX2 or above".
We should also update the copyright year.

Comment (shipilev) [reply src/hotspot/cpu/x86/macroAssembler_x86.hpp]: Rephrased the assert message, updated copyright date. Thanks!

**Review by vnkozlov — APPROVED** [review]
Good.

**Review by DamonFool — APPROVED** [review]
LGTM
Thanks for your update.

**Comment by shipilev** [comment]
Thank you for reviews!

/integrate

**Pull request closed without merging**